### DISCUSSION TITLE
MultiGPU/multinode support?

### DISCUSSION THREAD

**Discussion by jon-chuang** [intel/llvm#5572] [Q&A]
What is SYCL's abilities to accomodate multigpu (e.g. NVLink) and multinode (e.g. Infiniband GPUDirect) setups?

**Comment by AlexeySachkov** [comment]
Hi @jon-chuang, sorry for such long delay,

[Section 3.5. The SYCL platform model](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:platformmodel) of SYCL 2020 specification says:

> A SYCL [context](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#context) is constructed, either directly by the user or implicitly when creating a [queue](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#queue), to hold all the runtime information required by the SYCL runtime and the [SYCL backend](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#backend) to operate on a device, or group of devices. When a group of devices can be grouped together on the same [context](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#context), they have some visibility of each other’s memory objects. The SYCL runtime can assume that memory is visible across all devices in the same context. Not all devices exposed from the same [platform](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#platform) can be grouped together in the same [context](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#context).

Additionally, [Section 3.7.1.1. SYCL backend resources managed by the SYCL application](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_sycl_backend_resources_managed_by_the_sycl_application) says:

> [Contexts](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#context): any [SYCL backend](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#backend) resource that is acquired by the user is attached to a [context](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#context). A context contains a collection of devices that the host can use and manages memory objects that can be shared between the devices. Devices belonging to the same context must be able to access each other’s global memory using some implementation-specific mechanism. A given context can only wrap devices owned by a single platform. A context is exposed to the user with a sycl::context object.

So, SYCL spec says that implementation can provide to you several devices, which could be grouped into the same context, which will guarantee that the memory is shared between them.

I won't be able to list all mechanisms which are specifically designed with multi-device setup in mind, but as an example, from [Section 3.8.3. SYCL memory consistency model](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:memoryconsistency):

> Memory consistency between the host and SYCL device(s), or different SYCL devices in the same context, can be guaranteed through synchronization in the host application as defined in [Section 3.9.8](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:synchronization). On SYCL devices supporting concurrent atomic accesses to USM allocations and acquire-release or sequentially consistent memory orderings, **cross-device memory consistency can be enforced through the use of [mem-fence](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#mem-fence) and atomic operations.**

For Level Zero backend we have the following document: [Considerations for programming to multi-tile and multi-card under Level-Zero backend](https://github.com/intel/llvm/blob/sycl/sycl/doc/MultiTileCardWithLevelZero.md).

Tagging @Naghasan to see if he can add more about CUDA backend and multi-device setups

**Reply by AerialMantis** [reply]
I can answer to this on the CUDA side.

For the CUDA backend we have begun implementing support for multi-GPU by introducing an extension in PI which exposes the CUDA peer-to-peer copy capabilities (https://github.com/intel/llvm/pull/4401) and using this we have been able to enable peer-to-peer copies between Nvidia GPUs via PCIe or NVLink.

However as @AlexeySachkov mentioned above the way SYCL provides this is via a shared SYCL context as devices in the same context must be able access each other's global memory. Unfortunately multi-device contexts are currently not supported in the CUDA backend (https://github.com/intel/llvm/issues/4381), but once this is supported we'll be able to enable peer-to-peer copies for Nvidia devices within a context.
